File: internal/expr/printify.go
package expr

import (
	"fmt"
	"strings"
)

type PrintifyVisitor struct{}

func (v PrintifyVisitor) String(expression Expr) string {
	return expression.Accept(v).(string)
}

func (v PrintifyVisitor) Print(expression Expr) {
	fmt.Println("Expr:", v.String(expression))
}

func (v PrintifyVisitor) VisitBinaryExpr(b *Binary) any {
	return v.parenthesize(b.Operator.Word, b.Left, b.Right)
}

func (v PrintifyVisitor) VisitGroupingExpr(g *Grouping) any {
	return v.parenthesize("group ", g.Expression)
}

func (v PrintifyVisitor) VisitLiteralExpr(l *Literal) any {
	if l.Value == "" {
		return "null"
	}

	return fmt.Sprintf("%v", l.Value)
}

func (v PrintifyVisitor) VisitUnaryExpr(u *Unary) any {
	return v.parenthesize(u.Operator.Word, u.Right)
}

func (v PrintifyVisitor) parenthesize(name string, exprs ...Expr) any {
	var syntax []string

	syntax = append(syntax, "(", name)
	for _, exp := range exprs {
		syntax = append(syntax, exp.Accept(v).(string))
	}
	syntax = append(syntax, ")")

	return strings.Join(syntax, " ")
}

File: src/expr/printify.rs
use super::binary::Binary;
use super::grouping::Grouping;
use super::literal::Literal;
use super::unary::Unary;
use super::{Expr, ExprVisitor};

pub struct Printify {}

impl Printify {
    fn parenthesize(&self, name: String, expressions: Vec<Box<dyn Expr>>) -> String {
        let mut syntax: Vec<String> = vec![];

        syntax.push(format!("( {}", name));
        for expr in expressions.iter() {
            syntax.push(expr.accept(self));
        }
        syntax.push(format!(" )"));

        return syntax.join(" ");
    }
}

impl ExprVisitor<String> for Printify {
    fn visit_binary_expr(&self, binary: &Binary) -> String {
        "binary".to_string()
    }
    fn visit_grouping_expr(&self, grouping: &Grouping) -> String {
        "grouping".to_string()
    }
    fn visit_literal_expr(&self, literal: &Literal) -> String {
        "literal".to_string()
    }
    fn visit_unary_expr(&self, unary: &Unary) -> String {
        let right = vec![unary.right];
        return self.parenthesize(unary.operator.word, right);
    }
}
